### ISSUE TITLE
Duplicate entries even though primary key is the same

### ISSUE THREAD

**Issue by batjo** [3lvis/Sync#447]
I'm trying to sync the following json:

```json
[  
   {  
      "id":1,
      "recent_tags":[  
         {  
            "id":10
         }
      ],
      "categories":[  
         {  
            "id":100,
            "defaultTag":{  
               "id":10
            }
         }
      ]
   }
]
```

In this example, a Website has tags and categories. Since there can be lots of tags, only the most recent ones are being send. Categories can have a defaultTag.

When I sync the above JSON I end up with 2 tags, both with id: 10. 

Btw I just became a Patreon, thanks for your open source libraries, they are awesome!

**Comment by 3lvis** [comment]
>Btw I just became a Patreon, thanks for your open source libraries, they are awesome!

Woah! Thanks for your support Bart :). You're awesome 🌹.

**Comment by 3lvis** [comment]
Hi Bart,

The issue here was that since your website can have many tags and this tag can have many websites then there should be a many-to-many relationship. After updating to this model the tests pass again.

**Comment by 3lvis** [comment]
<img width="546" alt="screen shot 2017-11-07 at 23 24 01" src="https://user-images.githubusercontent.com/1088217/32521201-e9715472-c412-11e7-9e5d-6dd4f1242357.png">

**Comment by batjo** [comment]
Hi 3lvis thanks for having a look. In my case each website has it's own tags and they are not shared between websites. So although a website can have many tags, a tag belongs to only 1 website.

Is there a specific reason why I need to model this has many-to-many?

**Comment by 3lvis** [comment]
Hi @batjo,

I remember this issue has to do with diffing and the fact that in order to syncronize tags between multiple tables then we need to use many-to-many, if we use many-to-one, when diffing the best way to ensure that we are only deleting tags that you want to delete is if we make an exclusive tag for the relationship.

**Comment by 3lvis** [comment]
But I guess in your case if you don't include a tag inside "recent_tags" then we are supposed to delete it? Or what was your logic for diffing in this case?

**Comment by 3lvis** [comment]
Reopening until we find a good alternative.

**Comment by batjo** [comment]
Yes in my case tags that aren't included in "recent_tags" should be deleted. I guess for now I just use a many-to-many relationship as I don't see any downsides to it other then that my core data model differs slightly from the web server model.

**Comment by donghd** [comment]
@3lvis , @batjo :  I have same problem when sync 1 file (total: 19 leads , 4 stages) with Leads - Stage (many-to-one  relationship), it creates many Stages with same ID (that expected to one with unique ID), now I changed to (many-to-many), but it only saved the last connection (19th lead-stage, in json file), and still create duplicate Stages (8 stages, expected 4 stages) with same ID. Anyway, I will check more and try some another way ... If you have any tips or suggest, please kindly let me know.
Thanks,

**Comment by batjo** [comment]
I think the main issue here is that Sync doesn't know the `Tag` belongs to a `Website`, when it's in a `Category`. So the frist tags gets synced correctly and belongs to a Website, while the second tag doesn't have a Website according to sync.

So in the database it looks like this:

id | website
-- |--------
10 | 1
10 | null

For now I fixed this by preprocessing all the json and assigning a website_id to each tag.

**Comment by donghd** [comment]
I am considering and will test as below, that may work
- Sync  4 stages firstly (new different json data)
- Sync 19 leads with core data's stage with json's stage_id to linked to 1 of 4 pre-created stages record, that mean Sync only create relationship, to avoid create new stage's records.
that may be same as your comments @batjo

**Comment by donghd** [comment]
I have "Invalid expression" when refer such as lead.stage, even I did Sync data model lead's stage with json data's stage_id value. How should I fix?
 @batjo : as you mention you sync Tag's website (with json data {website_id: value}), can you access data model Website from Tag such as : Tag.website as Website  ??

**Comment by batjo** [comment]
Yes I get the complete website object with tag.website

**Comment by donghd** [comment]
Yes, I can see it work in simple case. However Sync doesn't recognize existing 'id', it always create new one with same 'id', even I will operations: [.insert, .update] . It is quite difficult incase we use parent_id to link to Model itself.

**Comment by donghd** [comment]
@3lvis : I would like to report the issue as below:
Task:
we need to sync Partner, with relationship to itself, to parent-partner(via json 'parent_id')
Issue:
1. It cannot sync correctly one time cause, parent_id's partner is not yet existed before, so it also save relationship 'parent' is null in core data.

2. So I separate into 2 time.
var update1: Store parent-partner (normally it has 'parent_id' = null)
var update1: Store child-partner (normally it has 'parent_id' <> null)
```
self.dataStack.sync(update1, inEntityNamed: ResPartner.entity().name!) { error in    
     self.dataStack.sync(update2, inEntityNamed: ResPartner.entity().name!, operations: [.insert, .update]) { error in                 
                    completion(.success(nil))
     }
}
```
1st time: OK (as expected)
2nd time: all objects of update2 are created new (have 2 objects same ID)
3rd: all objects of update2 are created new (have 3 objects same ID)
....
Can you check it?
Thanks

**Comment by batjo** [comment]
Hi @donghd, I think I fixed your issue, please check both these pull requests (you need them both) and let me know if that solved your issue:  #471 #459
